### ISSUE TITLE
Language tags: POSIX or BCP-47?

### ISSUE THREAD

**Issue by mebeim** [tldr-pages/tldr#3183]
This issue just popped up from [this thread on the `tldr-node-client` repo](https://github.com/tldr-pages/tldr-node-client/pull/270#discussion_r300885553), from which I'm quoting the relevant comments:

> @mebeim: Our client specification does not follow [the POSIX locale specification](https://www.gnu.org/software/gettext/manual/html_node/Locale-Names.html#Locale-Names) for language environment variables. As a result of this, the `LANG` environment var is basically never going to be right, and should be converted first. 
> [...]
> This is an issue I've been thinking about for quite a while. I really think we should update the client spec to follow the POSIX locale specification (basically just use `ll_CC` instead of `ll-CC`). It would make it really easier for clients to deal with languages.

> @pxgamer: I was thinking this earlier too. I know there was a reason that we chose the current language code format though, can't remember why.

> @mebeim: the only issue here is the dash - versus the underscore _. I don't remember a real reason for this choice, if I recall right 🤔.

> @agnivade: It's here - https://gitter.im/tldr-pages/tldr?at=5c387be59289cc7aa757c1ad. You commented yourself @mebeim :) 
>
> tldr is GNU uses _ and BCP-47 uses -. They have different standards. And we decided to go with BCP-47.

> @mebeim: Thanks for the link @agnivade makes a lot more sense now, we were looking at BCP-47 and not POSIX. That's a bummer since now that I think about it the POSIX version is probably the only one you'll find used around when it comes to programming and environment variables.

---

Currently TLDR pages uses a simplified version of the [IETF BCP 47 language tags](https://tools.ietf.org/rfc/bcp/bcp47.txt) (we only use the language and country fields). This is reflected in the client specification. In reality though, BCP 47 tags are quite uncommon, and on any POSIX compliant system [POSIX language tags](https://www.gnu.org/software/gettext/manual/html_node/Locale-Names.html#Locale-Names) are used instead. 

To put it simply, what all of this boils down to is the following:

 - We are currently using BCP 47 `ll-CC` tags (e.g. `pages.pt-BR`).
 - Most (if not all) operating systems use POSIX `ll_CC` tags. 

While the only thing that changes is a dash (`-`) versus an underscore (`_`), this seems to makes things unnecessarily more complicated for clients and in general any programmatic operation that has to deal with language tags.

I am therefore opening this issue to suggest switching to POSIX tags now that we can since that virtually no client is supporting languages already, and this might be the last chance to make such a change without much trouble.

If we choose to go with this, we would have to:

1. Rename pages folders (currently only `pages.pt-BR`).
2. Update the "Translations" section of the contributing guide.
3. Update the "Language" section of the client specification.
4. Deprecate any previous client specification (currently only two versions) version since specification-wise this is a breaking change. For this point I'd suggest to simply add `(deprecated)` alongside previous client spec versions in the "Changelog" section of the document.

I'd suggest the changes to the client spec into #3168 to limit the number of different versions of the spec.

What do you guys think?

**Comment by pxgamer** [comment]
I'd say we change to POSIX :+1: Easier to do it now, a simple change to the spec, and it should work out of the box with the clients without having to set the `LANG` env variable each time.

**Comment by agnivade** [comment]
_sigh_ .. It's hard to accept that we decided on the wrong thing. :disappointed:  But it's never too late.

I have a question for Windows though. What is the usual convention for command line tools to pick up language preferences ? I know there is a setting in the control panel, but does that follow any standard ?

**Comment by waldyrious** [comment]
> It's hard to accept that we decided on the wrong thing. 

I wouldn't say it's wrong — BCP 47 is a valid standard. The POSIX one is probably a _better choice_, as it's more prevalent in the environments we operate in, but the previous decision was perfectly legitimate. ;)

**Comment by agnivade** [comment]
I mean I didn't think hard enough that POSIX would be more common in the environment we operate in and hence is a better choice.

**Comment by andrik** [comment]
It's a miracle :smile: , such a change will not break the clients. Let's be honest, for software development this is a rare scenario.

I think it's the way to go

**Comment by mebeim** [comment]
Good question @agnivade, Windows does not have environment variables for language, it should be obtained through APIs like [`GetSystemDefaultUILanguage`](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getsystemdefaultuilanguage) or [`GetSystemPreferredUILanguages`](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getsystempreferreduilanguages). As per the language identifier itself,  Windows seems to use integer bitmasks for these APIs, but there also seem to be BCP47 style constants defined, like [`LOCALE_SNAME`](https://docs.microsoft.com/en-us/windows/win32/intl/locale-names). I don't know much about which one should be used and how though.

**Comment by sbrl** [comment]
Hrm. Windows sounds like it complicates everything - especially since it's no longer POSIX compliant. I wonder if an msys environment has the `LANG` environment variable?

I'd say we update the spec to use the POSIX spec instead, perhaps making a note saying that the language should be determined whatever system APIs are available if `LANG` doesn't exist (e.g. `GetSystemDefaultUILanguage()`,  `navigator.language`, etc.).

**Comment by pxgamer** [comment]
Pretty sure that MSYS, WSL, and Bash subsystem all have the `LANG` environment set.
![image](https://user-images.githubusercontent.com/1899334/60978046-6af57000-a328-11e9-9743-774b6193feb6.png)

**Comment by agnivade** [comment]
@mebeim - Do you want to take point on this ?
